genie-agent/src/main/java/com/netflix/genie/agent/cli/ResolveJobSpecCommand.java

```diff
@@ -0,0 +1,180 @@
+
+    private final ResolveJobSpecCommandArguments resolveJobSpecCommandArguments;
+    private final AgentJobSpecificationService agentJobSpecificationService;
+    private final JobRequestConverter jobRequestConverter;
+
+    ResolveJobSpecCommand(
+        final ResolveJobSpecCommandArguments resolveJobSpecCommandArguments,
+        final AgentJobSpecificationService agentJobSpecificationService,
+        final JobRequestConverter jobRequestConverter
+    ) {
+        this.resolveJobSpecCommandArguments = resolveJobSpecCommandArguments;
+        this.agentJobSpecificationService = agentJobSpecificationService;
+        this.jobRequestConverter = jobRequestConverter;
+    }
+
+    @Override
+    public void run() {
+        log.info("Resolving job specification");
+
+        final ObjectMapper prettyJsonMapper = GenieObjectMapper.getMapper()
+            .copy() // Don't reconfigure the shared mapper
+            .enable(SerializationFeature.INDENT_OUTPUT);
+
+        final JobSpecification spec;
+        final String specId = resolveJobSpecCommandArguments.getSpecificationId();
+        if (!StringUtils.isBlank(specId)) {
+            // Do a specification lookup if an id is given
+            try {
+                spec = agentJobSpecificationService.getJobSpecification(specId);
+            } catch (final JobSpecificationResolutionException e) {
+                throw new RuntimeException("Failed to get spec: " + specId, e);
+            }
+
+        } else {
+            // Compose a job request from argument
+            final AgentJobRequest agentJobRequest =
+                jobRequestConverter.agentJobRequestArgsToDTO(resolveJobSpecCommandArguments.getJobRequestArguments());
+
+            // Print request
+            if (!resolveJobSpecCommandArguments.isPrintRequestDisabled()) {
+                try {
+                    System.out.println(prettyJsonMapper.writeValueAsString(agentJobRequest));
+                } catch (final JsonProcessingException e) {
+                    throw new RuntimeException("Failed to map request to JSON", e);
+                }
+            }
+
+            // Resolve via service
+            try {
+                spec = agentJobSpecificationService.resolveJobSpecification(agentJobRequest);
+            } catch (final JobSpecificationResolutionException e) {
+                throw new RuntimeException("Failed to resolve job specification", e);
+            }
+        }
+
+        // Translate response to JSON
+        final String specJsonString;
+        try {
+            specJsonString = prettyJsonMapper.writeValueAsString(spec);
+        } catch (final JsonProcessingException e) {
+            throw new RuntimeException("Failed to map specification to JSON", e);
+        }
+
+        // Print specification
+        System.out.println(specJsonString);
+
+        // Write specification to file
+        final File outputFile = resolveJobSpecCommandArguments.getOutputFile();
```

Comment:
Might be cleaner here to have your `getOutputFile` return an `Optional` then you can do something like `resolveJobSpecCommandArguments.getOutputFile().ifPresent({})`... it flows slightly better (IMO) than writing it to a variable checking for null in an if block then doing the logic... but that's just me

Reply:
If i make the args delegate of type optional, then i need a custom "converter" for each of them.
I.e. something that goes from a `null` string to `Optional.ofNullable(null)`. I'd rather avoid. Args can be null and that's part of the contract.